mcorpc/authz_rego.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/choria-io/go-config"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/topdown"
+	"github.com/sirupsen/logrus"
+)
+
+type regoPolicy struct {
+	cfg   *config.Config
+	req   *Request
+	agent *Agent
+	log   *logrus.Entry
+}
+
+func regoPolicyAuthorize(req *Request, agent *Agent, log *logrus.Entry) (bool, error) {
+	logger := log.WithFields(logrus.Fields{
+		"authorizer": "regoPolicy",
+		"agent":      agent.Name(),
+		"request":    req.RequestID,
+	})
+
+	authz := &regoPolicy{
+		cfg:   agent.Config,
+		req:   req,
+		agent: agent,
+		log:   logger,
+	}
+
+	return authz.authorize()
+}
+
+func (r *regoPolicy) authorize() (bool, error) {
+	policyFile, err := r.lookupPolicyFile()
+	if err != nil {
+		return false, err
+	}
+
+	pol, err := ioutil.ReadFile(policyFile)
+	if err != nil {
+		return false, err
+	}
+
+	module := string(pol)
+
+	buf := topdown.NewBufferTracer()
+
+	options := []func(*rego.Rego){
+		rego.Query("data.choria.mcorpc.authpolicy.allow"),
+		rego.Module("//choria/mcorpc/policies/"+r.agent.meta.Name, module),
+		rego.Input(r.regoInputs()),
+	}
+
+	if r.log.Logger.GetLevel() == logrus.TraceLevel {
+		options = append(options, rego.Tracer(buf))
+	}
+
+	query := rego.New(
+		options...,
+	)
+
+	rs, err := query.Eval(context.Background())
+
+	if r.log.Logger.GetLevel() == logrus.TraceLevel {
+		topdown.PrettyTrace(r.log.Writer(), *buf)
+	}
+
+	switch err := err.(type) {
+	case nil:
+		break
+	case ast.Errors:
+		for _, e := range err {
+			r.log.Info("code: ", e.Code)
+			r.log.Info("row: ", e.Location.Row)
+			r.log.Info("filename: ", policyFile)
+		}
+	default:
+		return false, err
+	}
+
+	return rs[0].Expressions[0].Value.(bool), nil
+}
+
+func (r *regoPolicy) lookupPolicyFile() (string, error) {
+	dir := filepath.Join(filepath.Dir(r.cfg.ConfigFile), "policies", "rego")
+
+	regoPolicy := filepath.Join(dir, r.agent.Name()+".rego")
+
+	r.log.Debugf("Looking up rego policy in %s", regoPolicy)
+	if choria.FileExist(regoPolicy) {
+		return regoPolicy, nil
+	}
+
+	defaultPolicy := filepath.Join(dir, "default.rego")
+	if choria.FileExist(defaultPolicy) {
+		return defaultPolicy, nil
+	}
+	return "", fmt.Errorf("no policy %s found for %s in %s", defaultPolicy, r.agent.Name(), dir)
+
+}
+
+func (r *regoPolicy) regoInputs() map[string]interface{} {
+	facts := map[string]interface{}{}
+
+	err := json.Unmarshal(r.agent.ServerInfoSource.Facts(), &facts)
+	if err != nil {
+		r.log.Errorf("unable to unmarshal facts?")
+	}
+
+	return map[string]interface{}{
+		"agent":         r.req.Agent,
+		"action":        r.req.Action,
+		"callerID":      r.req.CallerID,
+		"collective":    r.req.Collective,
+		"ttl":           r.req.TTL,
+		"time":          r.req.Time,
+		"facts":         facts,
+		"classes":       r.agent.ServerInfoSource.Classes(),
+		"agents":        r.agent.ServerInfoSource.KnownAgents(),
+		"provisionMode": r.agent.Choria.ProvisionMode(),
```

Comment:
`r.agent.ServerInfoSource.Provisioning()`, i want to remove access to the choria instance from agents

Reply:
That's not defined in go-choria/server/agents/agents.go?

Reply:
ServerInfoSource is the whole instance of server.Instance, so all this stuff https://github.com/choria-io/go-choria/blob/master/server/infosource.go - and infact the whole server but the interface limits it.

Reply:
I can't figure out how to enable this.  I stepped through the tests with a debugger and unless the build time provisioning URLs/etc are set, it won't enable?

Reply:
`build.ProvisionBrokerURLs = "nats://localhost:4222/"` that should be enough